### PR TITLE
Associate PublishableEntities with Collections [FC-0062]

### PR DESCRIPTION
**Pull request by pomegranited**
## Description

Adds models and APIs for linking objects (PublishableEntities) to Collections, and for fetching the Collections associated with a given entity.

## Supporting information

Part of : https://github.com/openedx/modular-learning/issues/227
Based on : https://github.com/openedx/openedx-learning/pull/131

Private-ref: [FAL-3784](https://tasks.opencraft.com/browse/FAL-3784)

## Testing instructions

See https://github.com/open-craft/edx-platform/pull/674

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.




#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@axim-engineering`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [inline openedx_learning/apps/authoring/collections/api.py:161]: This method is out of scope for this ticket, but will be needed by https://github.com/openedx/modular-learning/issues/230

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [inline openedx_learning/apps/authoring/collections/api.py:151]: Hmm.. I thought we'd need this for https://github.com/openedx/modular-learning/issues/230 too, but now I'm unsure.. might only need `get_object_collections` below, since we can search for a Collection's components with a filter for `collection='<id>'`.

**Comment by bradenmacdonald** [comment]
@ormsbee Can you take a quick look at this when you get some time? In particular, if possible, I want to make sure that this is sufficiently general that we can link **units/sections/subsections** to collections with the same API, even if they're not implemented as XBlocks/Components. Is `PublishableEntity` the right "thing" to use here then?

**Comment by ormsbee** [comment]
@bradenmacdonald: I'll do a full review now this afternoon, but yes, `PublishableEntity` is the correct generic thing to join against to get all the different publishable things.

File: openedx_learning/apps/authoring/collections/api.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/collections/api.py:35]: Please change to `entities_qset` for consistency with publishing app models.

**Review by ormsbee — CHANGES_REQUESTED** [review]
Some questions and requests. Most of these are minor changes to match `openedx_learning` package coding conventions that I have failed to document anywhere, and there is no reasonable way for you to have been aware of them. 😛 Sorry about that, and thank you for your patience.

**Review by rpenido — APPROVED** [review]
Hi @pomegranited!

Thank you for your work here! I just left some comments.

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py:54]: @ormsbee I don't have a use case for creating a collection with components in it, I was just copying that code from your PR. Maybe it's better just to keep them as separate API calls?

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py:91]: I wouldn't think so, but that's what was in the ticket: https://github.com/openedx/modular-learning/issues/227

@bradenmacdonald can you comment?

File: openedx_learning/apps/authoring/collections/models.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/models.py:146]: Sure thing.

FYI @rpenido I'm going to replace the database migration in this PR, so here's how to roll it back in your devstack:

```
openedx-learning $ git checkout 0e15dd0d6445464b73eb8572c8bc83f8b0fdee16
(tutor) $ tutor dev run cms bash
app@xxx:~/edx-platform$ ./manage.py lms migrate oel_collections 0002_remove_collection_name_collection_created_by_and_more
```

File: openedx_learning/apps/authoring/collections/models.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/models.py:10]: Oh definitely.. @ormsbee 's notes are always worth preserving :)

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py:124]: `timezone.now()` doesn't accept a `tz` parameter, but the [Django docs](https://docs.djangoproject.com/en/4.2/ref/utils/#django.utils.timezone.now) say:

> If [USE_TZ](https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-USE_TZ) is True, this will be an [aware](https://docs.djangoproject.com/en/4.2/topics/i18n/timezones/#naive-vs-aware-datetimes) datetime representing the current time in UTC. Note that [now()](https://docs.djangoproject.com/en/4.2/ref/utils/#django.utils.timezone.now) will always return times in UTC regardless of the value of [TIME_ZONE](https://docs.djangoproject.com/en/4.2/ref/settings/#std-setting-TIME_ZONE);

If someone has un-set `settings.USE_TZ`, then they'll probably have all kinds of problems, so I think this is ok as-is.

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [inline openedx_learning/apps/authoring/collections/api.py:181]: TODO: rename `get_learning_package_collections` to `get_collections`, cf https://github.com/openedx/edx-platform/pull/35321#discussion_r1734711274

**Review by pomegranited — COMMENTED** [review]
Thank you for your reviews, @ormsbee and @rpenido :) I addressed everything I could, but there's a couple open questions, if you could take another look?

File: openedx_learning/apps/authoring/collections/api.py
Comment (rpenido) [inline openedx_learning/apps/authoring/collections/api.py:134]: :+1:

File: openedx_learning/apps/authoring/collections/api.py
Comment (rpenido) [inline openedx_learning/apps/authoring/collections/api.py:184]:
```suggestion
    Only enabled collections are returned by default
```

**Review by rpenido — APPROVED** [review]
Thank you for the changes, @pomegranited!
LGTM :+1:

File: openedx_learning/apps/authoring/collections/api.py
Comment (bradenmacdonald) [reply openedx_learning/apps/authoring/collections/api.py:92]: Sorry, I was just trying to indicate that the API needs to support having an item linked to multiple collections, but it doesn't necessarily have to support linking multiple collections to an item in a single API call. If it simplifies the implementation to only operate by adding/removing a single collection at a time, that's fine with me.

There is only one use case I can think of where we might want to associate/dissociate from multiple collections simultaneously and that's import/export. But we haven't planned that part out yet at all.

File: openedx_learning/apps/authoring/collections/api.py
Comment (bradenmacdonald) [reply openedx_learning/apps/authoring/collections/api.py:127]: Other parts of the code use `datetime.now(tz=timezone.utc)` - so I think you can just use that.

File: openedx_learning/apps/authoring/collections/api.py
Comment (ormsbee) [reply openedx_learning/apps/authoring/collections/api.py]: I'm fine either way. I think that use case will come up when we do v1 library import functionality, but I'm fine with punting until then.

File: openedx_learning/apps/authoring/collections/api.py
Comment (ormsbee) [reply openedx_learning/apps/authoring/collections/api.py:127]: @pomegranited: I think that you're right that it will probably break a bunch of things if they don't have `USE_TZ=True`. I hadn't realized that it would always return UTC timezones if that setting was on–I had assumed it would return whatever the configured `TIME_ZONE` setting was set to.

As @bradenmacdonald notes, I've always just used `datetime.now(tz=timezone.utc)`.

So I have a mild preference for `datetime.now(tz=timezone.utc)`, but I'm fine with either.

File: openedx_learning/apps/authoring/collections/api.py
Comment (ormsbee) [reply openedx_learning/apps/authoring/collections/api.py:129]: Oh, that's a fantastic callout. We should absolutely enforce it at the Learning Core layer, so people don't shoot themselves in the foot. It's very possible we'll eventually allow folks to add stuff from other LearningPackages into Collections, but we're going to have to be very deliberate when we do and think through the lifecycle issues when those LearningPackages disappear. We definitely don't want to let it happen by accident because other apps used Learning Core directly instead of going through the Libraries REST API.

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py]: I ended up [removing the `entities_qset` param from `create_collection`](https://github.com/openedx/openedx-learning/pull/216/commits/2fe2e708fa208a0cfee33805afaa3089f7d70881#diff-91f68d06863987ec50029b7bcdb05b4ff2edde0052aca94b79e3dbd2e0c5c742R41-R48), we can add it back later if needed.

**Comment by pomegranited** [comment]
Thanks for your review @rpenido :)  How's this looking now, @ormsbee ?

File: openedx_learning/apps/authoring/collections/api.py
Comment (ormsbee) [inline openedx_learning/apps/authoring/collections/api.py:141]: [speculation (please do not block PR for this)]: It might be nice if we could a custom Manager on the through-model so that if someone has an `entity`, they could do something like `entity.collections.active()`.

**Review by ormsbee — APPROVED** [review]
Optional nitpick and some speculation, but nothing that would block merging. Thank you!

(Also, this is a three-day weekend in the U.S., so I may be a bit slow on reviews.)

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py:141]: Oh absolutely -- I am interested to see the other use cases for Collections besides what we're supporting here. But will leave that for a day when it's needed, thank you :)

File: openedx_learning/apps/authoring/collections/api.py
Comment (pomegranited) [reply openedx_learning/apps/authoring/collections/api.py]: That's much cleaner, thank you, done with [e03daaa](https://github.com/openedx/openedx-learning/pull/216/commits/e03daaa36b07afd48197ed9b3940686a6e05d206)

**Comment by openedx-webhooks** [comment]
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-learning&entry.867055334=https://github.com/openedx/openedx-learning/pull/216&entry.1484655318=https://github.com/pomegranited&entry.752974735=2024-08-21+07:40&entry.1917517419=2024-09-03+00:31&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
